dns/route53/dns_test.go

```diff
@@ -0,0 +1,166 @@
+package route53
+
+import (
+	"testing"
+	"time"
+
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+
+const (
+	MOCKED_ZONE_ID = "20N31D"
+)
+
+
+type mockR53 struct {
+	recordSets map[string]*route53.ResourceRecordSet
+}
+
+func (r53 *mockR53) ChangeResourceRecordSets(input *route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error) {
+	for _, change := range input.ChangeBatch.Changes {
+		switch *change.Action {
+		case route53.ChangeActionCreate, route53.ChangeActionUpsert:
+			r53.recordSets[*change.ResourceRecordSet.Name] = change.ResourceRecordSet
+
+		case route53.ChangeActionDelete:
+			delete(r53.recordSets, *change.ResourceRecordSet.Name)
+
+		default:
+			panic(fmt.Sprintf("Invalid action: %s", *change.Action))
+		}
+	}
+
+	return &route53.ChangeResourceRecordSetsOutput{
+		ChangeInfo: &route53.ChangeInfo{
+			Id:          aws.String("MOCKED-AWS-API-ID"),
+			Status:      aws.String("INSYNC"),
+			SubmittedAt: aws.Time(time.Now()),
+		},
+	}, nil
+}
+
+func (r53 *mockR53) ListResourceRecordSets(input *route53.ListResourceRecordSetsInput) (*route53.ListResourceRecordSetsOutput, error) {
+	recordName := *input.StartRecordName
+
+	resourceRecordSets := []*route53.ResourceRecordSet{}
+	if r53.recordSets[recordName] != nil {
+		resourceRecordSets = append(resourceRecordSets, r53.recordSets[recordName])
+	}
+
+	return &route53.ListResourceRecordSetsOutput{
+		ResourceRecordSets: resourceRecordSets,
+		IsTruncated:        aws.Bool(false),
+		MaxItems:           input.MaxItems,
+	}, nil
+}
+
+// NOTE: always returns a list of one item matching the queried DNSName
+func (r53 *mockR53) ListHostedZonesByName(input *route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error) {
+	return &route53.ListHostedZonesByNameOutput{
+		HostedZones: []*route53.HostedZone{
+			&route53.HostedZone{
+				Id:              aws.String(MOCKED_ZONE_ID),
+				CallerReference: aws.String("MOCKED-CAL-REF"),
+				Name:            input.DNSName,
+			},
+		},
+		IsTruncated: aws.Bool(false),
+		MaxItems:    input.MaxItems,
+	}, nil
+}
+
+
```

Comment:
This should have a comment explaining what's tested - the removal of the entire record if there was only one record in the set.

Also, given this is very much a duplicate of `TestIntegrationAddAndRemove`, we should dedup the code running `srvManager.Add/Delete` and the assertions into a function where we only pass in an instantiated manager.

Reply:
I dont think de-duplicating the code is desired. `go test` caches the test (results) by default. Which meant that at least the integration tests can behave differently per run based on whether certain environment variables are set or not. 
Furthermore, these different types of tests - specifically the actual test logic - might diverge over time. Apart form this, to me, it doesnt sound like a good idea, to introduce a common dependency to different test types.

Reply:
> This should have a comment explaining what's tested

Renamed function to be more meaningful in terms of what it is testing.